internal/selecthost/selector.go

```diff
@@ -0,0 +1,501 @@
+
+		if current == nil {
+			span.AddEvent("selected host: " + selected.ID())
+
+			s.logger.Infow("Host Selected",
+				"selected.host", selected.ID(),
+				"selected.check_duration_ms", toMilliseconds(selected.AverageDuration()),
+			)
+		} else {
+			span.AddEvent("host changed: " + current.ID() + " -> " + selected.ID())
+
+			s.logger.Warnw("Host Changed",
+				"previous.host", current.ID(),
+				"previous.check_duration_ms", toMilliseconds(current.AverageDuration()),
+				"selected.host", selected.ID(),
+				"selected.check_duration_ms", toMilliseconds(selected.AverageDuration()),
+			)
+		}
+	} else if selected != nil && current == selected && !sticky {
+		// If host remains the same but stickiness has expired, reset the sticky counter.
+		s.stickyUntil = time.Now().Add(5 * s.checkInterval)
+	}
+
+	if selected == nil {
+		currentID := ""
+		if current != nil {
+			currentID = current.ID()
+		}
+
+		span.SetStatus(codes.Error, "unable to select host")
+
+		s.logger.Errorw("Unable to select host", "selected.host", currentID)
+	}
+}
+
+func (s *Selector) discovery() {
+	ticker := time.NewTicker(s.discoveryInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-s.runCh:
+			return
+		case <-ticker.C:
+		}
+
+		s.discoverRecords(context.Background())
+	}
+}
+
+func (s *Selector) discoverRecords(ctx context.Context) {
+	target := s.target
+
+	if strings.Contains(target, ":") {
+		target, _, _ = net.SplitHostPort(target)
+	}
+
+	ctx, span := tracer.Start(ctx, "discoverRecords", trace.WithAttributes(
+		attribute.String("discover.target", target),
+		attribute.String("discover.service", s.service),
+		attribute.String("discover.protocol", s.protocol),
+	))
+	defer span.End()
+
+	origCtx := ctx
+
+	ctx, cancel := context.WithTimeout(ctx, s.discoveryTimeout)
+	defer cancel()
+
+	start := time.Now()
+
+	logger := s.logger.With(
+		"discover.target", target,
+		"discover.service", s.service,
+		"discover.protocol", s.protocol,
+	)
+
+	logger.Debugf("Looking for srv records for service '%s' with protocol '%s' for target '%s'", s.service, s.protocol, target)
+
+	cname, srvs, err := s.resolver.LookupSRV(ctx, s.service, s.protocol, target)
+
+	span.SetAttributes(attribute.String("discover.resolved.cname", cname))
+
+	duration := time.Since(start)
+
+	logger = logger.With(
+		"discover.resolved.cname", cname,
+		"discover.runtime_ms", toMilliseconds(duration),
+	)
+
+	if err != nil {
+		span.RecordError(err)
+
+		if dnsErr, ok := err.(*net.DNSError); ok && dnsErr.IsNotFound && s.optional {
+			s.optionalFailureOnce.Do(func() {
+				span.AddEvent("no srv records found, however records are optional")
+
+				logger.Warnw("No SRV records found, using target/fallback.")
+			})
+		} else {
+			span.SetStatus(codes.Error, "Failed to lookup SRV records: "+err.Error())
+
+			logger.Errorw("Failed to lookup SRV records", "error", err)
+		}
+	}
+
+	s.mu.Lock()
+
+	added, removed, matched := diffHosts(s, srvs)
```

Comment:
If `s.resolver.LookupSRV` fails, we mark the span with an error but we still end up proceeding and using the srvs value. Should we return if we couldn't get the records?